README.md
# Download Synology Archive

<a href="https://github.com/007revad/Download_Synology_Archive/releases"><img src="https://img.shields.io/github/release/007revad/Download_Synology_Archive.svg"></a>
<a href="https://hits.seeyoufarm.com"><img src="https://hits.seeyoufarm.com/api/count/incr/badge.svg?url=https%3A%2F%2Fgithub.com%2F007revad%2FDownload_Synology_Archive&count_bg=%2379C83D&title_bg=%23555555&icon=&icon_color=%23E7E7E7&title=views&edge_flat=false"/></a>
[![](https://img.shields.io/static/v1?label=Sponsor&message=%E2%9D%A4&logo=GitHub&color=%23fe8e86)](https://github.com/sponsors/007revad)
[![committers.top badge](https://user-badge.committers.top/australia/007revad.svg)](https://user-badge.committers.top/australia/007revad)
<!-- [![committers.top badge](https://user-badge.committers.top/australia_public/007revad.svg)](https://user-badge.committers.top/australia_public/007revad) -->
<!-- [![committers.top badge](https://user-badge.committers.top/australia_private/007revad.svg)](https://user-badge.committers.top/australia_private/007revad) -->
<!-- [![Github Releases](https://img.shields.io/github/downloads/007revad/download_synology_archive/total.svg)](https://github.com/007revad/Download_Synology_Archive/releases) -->

### Description

Easily download part or all of the Synology download archive

### Main php script

You can run syno_archive_clone.php directly. It needs 2 or 3 arguments.

```
Usage: clone.php <srcdir> <subdir> [arch]
  Examples:
    php ./clone.php Os DSM
    php ./clone.php Package Docker
    php ./clone.php ToolChain toolkit

To download all in the <srcdir> use All as the <subdir>. ToolChain does NOT support the All argument
  Examples:
    php ./clone.php Os All
    php ./clone.php Package All

Firmware needs to include the firmware type in <scrdir>
  Examples:
    php ./clone.php Firmware/Camera BC500
    php ./clone.php Firmware/Camera TC500
    php ./clone.php Firmware/Camera All

"Synology NAS GPL Source" needs to be included in <scrdir> and CPU arch needs to be included as [arch]
  Examples:
    php ./clone.php ToolChain "Synology NAS GPL Source" v1000
    php ./clone.php ToolChain "Synology NAS GPL Source" geminilake
```

**Logging the output**

If you want to log the output with the date and time:

```
php ./clone.php <srcdir> <subdir> 2>&1 | tee ./clone_$(date '+%Y%m%d-%H%M').log

Example:
php ./clone.php Os DSM 2>&1 | tee ./clone_$(date '+%Y%m%d-%H%M').log
```

**Requirements**

There is 1 variable you need to set so the script knows where you want it to download to:

    $destination = "/volume1/downloads/archive.synology.com";

**Credit**

clone.php is based on the excellent work at [github.com/stopforumspam/download-synology-dsm](https://github.com/stopforumspam/download-synology-dsm) <br>
Give them a star if you find this script useful or interesting.

<br>

### Bash launcher script

To make it easier to run clone.php there is also syno_archive_clone.sh which provides menus to select the available `<srcdir>` and `<subdir>` options. It then runs the php script with those options as arguments.

To run syno_archive_clone.sh

    ./syno_archive_clone.sh

**Requirements**

There are 3 variables you need to set so the script knows where the 2 php scripts are and where to save the log:

    php_script="/volume1/scripts/get_urls.php"
    
    php_script="/volume1/scripts/clone.php"
    
    php_log="/volume1/downloads/archive.synology.com/clone_$(date '+%Y%m%d-%H%M').log"

php_log can be blank `php_log=""` if you don't want anything logged. Or you can disable it by putting a # in front of it:

    #php_log="/volume1/downloads/archive.synology.com/clone_$(date '+%Y%m%d-%H%M').log"

<br>

**Bash launcher script screenshots**

<p align="leftr"><img src="images/os.png"></p>

<p align="leftr"><img src="images/os-all.png"></p>

<p align="leftr"><img src="images/package-p1.png"></p>

<p align="leftr"><img src="images/package-p2.png"></p>

<p align="leftr"><img src="images/utility.png"></p>

<p align="leftr"><img src="images/mobile.png"></p>

<p align="leftr"><img src="images/chromeapp.png"></p>

<p align="leftr"><img src="images/source.png"></p>

<p align="leftr"><img src="images/firmware-2.png"></p>


CHANGES.txt
v1.0.5
- Now working for /download/Firmware/Camera

v1.0.4
- Now gets list of download types and sub folders instead of using a hard-coded list.
- Checks required files exist are are executable.
- Checks log path exists.
- Added timeout error handling for file_get_contents function.
- Known issue: Currently not working for /download/Firmware/Camera

v1.0.3
- Added arguments to download different types.
- Added a bash launcher script to make selecting the arguments easy.


get_urls.php
<?php
//--------------------------------------------------------------------
// Forked from https://github.com/stopforumspam/download-synology-dsm
//--------------------------------------------------------------------
// https://github.com/007revad/Download_Synology_Archive
//
// Required script for syno_archive_clone.sh
//--------------------------------------------------------------------

if (isset($argv[1])) {
    //echo "$argv[1]" . "\n";  // debug
    $srcdir = "$argv[1]";
}

function getDirs($url, $srcdir) {
    $html = file_get_contents($url);
    $dom = new DOMDocument;
    @$dom->loadHTML($html);
    foreach ($dom->getElementsByTagName('a') as $node) {

        $remote = parse_url($node->getAttribute("href"));
        $fullpath = explode("/", $remote["path"]);
        $type = urldecode(array_pop($fullpath));

        if (($type != "download") && (!str_contains($srcdir, $type))) {
            echo "\"" . $type . "\" ";  # Output to bash script
            $path = "download/$srcdir/$type/";
        }
    }
}

if ( (isset($srcdir)) && (!empty($srcdir)) ) {
    getDirs("https://archive.synology.com/download/$srcdir", "$srcdir");
} else {
    getDirs("https://archive.synology.com/download", "");
}
echo "\n";

exit();



syno_archive_clone.sh
#!/usr/bin/env bash
#--------------------------------------------------------------------
# Forked from https://github.com/stopforumspam/download-synology-dsm
#--------------------------------------------------------------------
# https://github.com/007revad/Download_Synology_Archive
#
# Checked with https://shellcheck.net/
#
# Script to select arguments for clone.php
#--------------------------------------------------------------------

# Set location of get_urls.php
php_urls="/volume1/scripts/get_urls.php"

# Set location of clone.php
php_script="/volume1/scripts/clone.php"

# Set location to save clone.php log
php_log="/volume1/downloads/archive.synology.com/clone_$(date '+%Y%m%d-%H%M').log"


#--------------------------------------------------------------------

echo -e "Synology Archive Clone \n"

check_file_exists() { 
    if [[ ! -f "$1" ]]; then
        echo -e "ERROR! File missing: \n$1"; exit 2
    elif [[ ! -x "$1" ]]; then
        echo -e "ERROR! File not executable: \n$1"; exit 13
    fi
}

check_file_exists "$php_urls"
check_file_exists "$php_script"

log_path=$(dirname -- "$php_log")
if [[ ! -d "$log_path" ]]; then
    echo -e "WARNING Log directory not found: \n$log_path"
    echo -e "Skipping logging \n"; php_log=""
fi

urls_to_array() { 
    num=2
    while [[ $num -lt $((qty +1)) ]]; do
        item=$(echo "$urls" | cut -d"\"" -f$num)
        #echo "$item"  # debug
        if [[ $item != "$srcdir" ]]; then
            typelist+=("$item")
        fi
        num=$((num +2))
    done
}

array_qty() { 
    qty=$(echo "$urls" | grep -o '"' | wc -l)  # Qty of " in string
    if [[ $1 != "NoAll" ]] && [[ $qty -gt "2" ]]; then
        typelist=( "All" )
    else
        typelist=()
    fi
}

# Get array of source directories
urls=$(php "$php_urls" "")
array_qty NoAll
urls_to_array

# Choose source directory
PS3="Choose download type: "
select srcdir in "${typelist[@]}"; do
    case "$srcdir" in
        "")
            echo "Invalid Choice!"
            ;;
        *)
            urls=$(php "$php_urls" "$srcdir")
            array_qty
            urls_to_array
            break
            ;;
    esac
done
echo -e "You selected: $srcdir \n"

# Choose sub directory
PS3="Choose $srcdir type: "
select subdir in "${typelist[@]}"; do
    case "$subdir" in
        "")
            echo "Invalid Choice!"
            ;;
        *)
            type="$subdir"
            break
            ;;
    esac
done
echo -e "You selected: $type \n"


# Choose Firmware sub directory
if [[ $srcdir = "Firmware" ]]; then
    srcdir="${srcdir}/$subdir"
    urls=$(php "$php_urls" "$srcdir")
    array_qty
    urls_to_array
    PS3="Choose $type type: "
    select subdir in "${typelist[@]}"; do
        case "$srcdir" in
            "")
                echo "Invalid Choice!"
                ;;
            *)
                type="$subdir"
                echo -e "You selected: $(basename -- "$srcdir") \n"
                break
                ;;
        esac
    done
fi

# Run clone.php
if [[ -n $srcdir ]] && [[ -n $type ]]; then
    if [[ -n $php_log ]]; then
        php "$php_script" "$srcdir" "$type" 2>&1 | tee "$php_log"
    else
        php "$php_script" "$srcdir" "$type"
    fi
else
    echo "args empty!"
    exit 1
fi

exit



syno_archive_clone.php
<?php
//-----------------------------------------------------------------------------
// Forked from https://github.com/stopforumspam/download-synology-dsm
//-----------------------------------------------------------------------------
// https://github.com/007revad/Download_Synology_Archive
//
// Checked at https://phpcodechecker.com/
//
// Run command
// php ./syno_archive_clone.php 2>&1 | tee ./clone_$(date '+%Y%m%d-%H%M').log
//-----------------------------------------------------------------------------

$destination = "/volume1/NetBackup/archive.synology.com";

if ($argc != "3") {
    if ($argc < "2") {
        echo "Missing arguments!" . "\n";
    }
    elseif ($argc < "3") {
        echo "Missing argument!" . "\n";
    }
    elseif ($argc > "3") {
        echo "Too many arguments!" . "\n";
    }

    $i = 1;
    $args = "$argv[0]";
    while($i < $argc) {
        $args = $args . " \"" . "$argv[$i]" . "\"";
        $i++;
    }
    echo $args . "\n\n";

    echo "Usage: " . "\n";
    echo "syno_archive_clone.php <srcdir> <subdir>" . "\n\n";
    echo "<subdir> can be All to download all in the srcdir>" . "\n\n";
    echo "Examples: " . "\n";
    echo "    php ./syno_archive_clone.php Os DSM" . "\n";
    echo "    php ./syno_archive_clone.php Os All" . "\n";
    echo "    php ./syno_archive_clone.php Package Docker" . "\n";
    echo "    php ./syno_archive_clone.php Package All" . "\n";
    echo "    php ./syno_archive_clone.php ToolChain toolkit" . "\n";
    echo "    php ./syno_archive_clone.php ToolChain \"Synology NAS GPL Source\"" . "\n";
    exit(1);
} else {
    $srcdir = "$argv[1]";
    $type = "$argv[2]";
    echo "------------------------------------------------------------ \n";
    //echo "Downloading: " . $srcdir . " > " . $type . "\n\n";
}

if(@chdir($destination)===true) { 
    echo "Current directory: " . getcwd() . "\n";
    sleep(5);  # Give me time to cancel if current directory incorrect
} else {
    echo "Change directory failed: " . $destination . "\n";
    exit(1);
}
echo "------------------------------------------------------------ \n";

function getLinks($url, $srcdir, $type, $dir) {

    $html = file_get_contents($url);
    $dom = new DOMDocument;
    @$dom->loadHTML($html);
    foreach ($dom->getElementsByTagName('a') as $node) {
        if (strpos($node->getAttribute("href"), ".pat") !== false  // Os
            || strpos($node->getAttribute("href"), ".spk") !== false  // Package
            || strpos($node->getAttribute("href"), ".bz2") !== false  // Utility
            || strpos($node->getAttribute("href"), ".deb") !== false  // Utility
            || strpos($node->getAttribute("href"), ".dmg") !== false  // Utility
            || strpos($node->getAttribute("href"), ".eap") !== false  // Utility
            || strpos($node->getAttribute("href"), ".elf") !== false  // Utility
            || strpos($node->getAttribute("href"), ".exe") !== false  // Utility
            || strpos($node->getAttribute("href"), ".gz") !== false   // Utility
            || strpos($node->getAttribute("href"), ".iso") !== false  // Utility
            || strpos($node->getAttribute("href"), ".msi") !== false  // Utility
            || strpos($node->getAttribute("href"), ".ova") !== false  // Utility
            || strpos($node->getAttribute("href"), ".pkg") !== false  // Utility
            || strpos($node->getAttribute("href"), ".rpm") !== false  // Utility
            || strpos($node->getAttribute("href"), ".vib") !== false  // Utility
            || strpos($node->getAttribute("href"), ".zip") !== false  // Utility
            || strpos($node->getAttribute("href"), ".apk") !== false  // Mobile
            || strpos($node->getAttribute("href"), ".crx") !== false  // ChromeApp
            || strpos($node->getAttribute("href"), ".txz") !== false  // ToolChain
            || strpos($node->getAttribute("href"), ".tbz") !== false  // ToolChain
            || strpos($node->getAttribute("href"), ".tgz") !== false  // ToolChain
            || strpos($node->getAttribute("href"), ".bin") !== false  // Firmware (Camera)
        ) {

            //echo "FILE: " . $node->getAttribute("href")."\n" ; 

            $remote = parse_url($node->getAttribute("href"));
            $fullpath = explode("/", $remote["path"]);
            $filename = urldecode(array_pop($fullpath));
            $path = "download/$srcdir/$type/$dir/";

            @mkdir($path, 0777, true);
            $dest = "$path$filename";

            // getting remote file size
            $ch = curl_init($node->getAttribute("href"));
            curl_setopt($ch, CURLOPT_NOBODY, true);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_HEADER, true);
            curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
            $data = curl_exec($ch);
            curl_close($ch);

            $contentLength = 'unknown';
            $status = 'unknown';
            if (preg_match('/^HTTP\/(1\.[01]|2) (\d\d\d)/', $data, $matches)) {
                 $status = (int)$matches[2];
            }
            if (preg_match('/Content-Length: (\d+)/i', $data, $matches)) {
                 $contentLength = (int)$matches[1];
            }

            if ($status == 200 && $contentLength <> "unknown" && file_exists($dest) && (filesize($dest) != $contentLength)) {
                echo "Remote file is $contentLength however local file is " . filesize($dest) . ", Redownloading\n";
                @unlink($dest);
            }

            if (!file_exists($dest)) {
                //echo "downloading " . $node->getAttribute("href") . " to $dest \n";
                echo "Downloading: " . "$dest \n";
                // echo "From: " . $node->getAttribute("href") . "\n\n";

                $fh = fopen($dest, "w");
                $ch = curl_init();
                curl_setopt($ch, CURLOPT_USERAGENT,"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/113.0 ");             
                curl_setopt($ch, CURLOPT_URL, $node->getAttribute("href"));
                curl_setopt($ch, CURLOPT_FILE, $fh);
                curl_exec($ch);
                curl_close($ch); 
//            } else {
//                // echo "Skipping download\n";
//                echo "Skipping: " . "$dest \n";
            }
        } else
        if (strpos($node->getAttribute("href"), "/download/$srcdir/$type/") !== false) {
            echo "DIR: " . $node->getAttribute("href")."\n";
            $url = "https://archive.synology.com" .$node->getAttribute("href");
            getLinks($url, $srcdir, $type, $node->nodeValue);
        }
    }
}

function getDirs($url, $srcdir, $type) {
    global $types;
    $types = array();

    $html = file_get_contents($url);
    $dom = new DOMDocument;
    @$dom->loadHTML($html);
    foreach ($dom->getElementsByTagName('a') as $node) {

        $remote = parse_url($node->getAttribute("href"));
        $fullpath = explode("/", $remote["path"]);
        $type = urldecode(array_pop($fullpath));

        if ($type != "download") {
            $types[] = $type;
            $path = "download/$srcdir/$type/";
        }
    }
}


$start = date('Y-m-d H:i:s');

if ($type != "All") {
    echo "Checking: " . $srcdir . "/" . $type . "\n";
    getLinks("https://archive.synology.com/download/$srcdir/$type", "$srcdir", "$type", "");
} else {
    echo "Checking: " . $srcdir . " > " . "All" . "\n";
    getDirs("https://archive.synology.com/download/$srcdir", $srcdir, "");
    $type = "";
    $i = 0;
    while($i < count($types)) {
        echo "\nChecking: " . $srcdir . "/" . $types[$i] . "\n";
        getLinks("https://archive.synology.com/download/$srcdir/$types[$i]", "$srcdir", "$types[$i]", "");
        $i++;
    }
}

echo "----------------------------------- \n";
echo "Started:  $start \n";
echo "Finished: " . date('Y-m-d H:i:s') . "\n";
echo "----------------------------------- \n";
 


syno_archive_select.sh
#!/usr/bin/env bash
#------------------------------------------------------------------------------
# Forked from https://github.com/stopforumspam/download-synology-dsm
#------------------------------------------------------------------------------
# https://github.com/007revad/Download_Synology_Archive
#
# Checked with https://shellcheck.net/
#
# Script to select arguments for syno_archive_clone.php
#------------------------------------------------------------------------------

php_script="/volume1/scripts/syno_archive_clone.php"
php_log="/volume1/downloads/archive.synology.com/clone_$(date '+%Y%m%d-%H%M').log"


echo -e "Synology Archive Clone \n"

PS3="Choose download type: "
select srcdir in "Os" "Package" "Utility" "Mobile" "ChromeApp" "ToolChain" "Firmware"; do
    case "$srcdir" in
        Os)
            typelist=( "All" "BSM" "DSM" "DSMUC" "SRM" "VS600HD" "VSF" "VSM" )
            break ;;
        Package)
            typelist=( "All" "AcronisTrueImage" "ActiveBackup" "ActiveBackup-GSuite" )
            typelist+=( "ActiveBackup-Office365" "ActiveDirectoryServer" "ActiveInsight" )
            typelist+=( "AntiVirus" "Apache2.2" "Apache2.4" "ArchiwareP5" "ArchiwarePure" )
            typelist+=( "AudioStation" "BackupRestoreManager" "BitDefenderForMailPlus" )
            typelist+=( "C2IdentityLDAPAgent" "Calendar" "CardDAVServer" "Chat" "CloudStation" )
            typelist+=( "CloudStationClient" "CloudSync" "CMS" "CodecPack" "Contacts" )
            typelist+=( "ContainerManager" "DdbBackup" "DhcpServer" "DHCPServer" )
            typelist+=( "DiagnosisTool" "DiagnosisTool-Eng" "DirectoryServer" )
            typelist+=( "DirectoryServerForWindowsDomain" "DNSServer" "Docker" "Docker-GitLab" )
            typelist+=( "Docker-LXQt" "Docker-Redmine" "DocumentViewer" "DokuWiki" "domotz" )
            typelist+=( "DownloadStation" "DVBLinkServer" "elephantdrive" "EmbyServer" )
            typelist+=( "Exchange-Migrate" "exFAT-Free" "EynioServer" "FileStation" )
            typelist+=( "geoip-database-city" "geoip-database-country" "Git" "GlacierBackup" )
            typelist+=( "GLPI" "GoodSync" "GoodSync Server" "GWS-Migrate" "HAManager" )
            typelist+=( "HighAvailability" "HybridShare" "HyperBackup" "HyperBackup-ED" )
            typelist+=( "HyperBackupVault" "IDrive" "iTunesServer" "Java8" "jdboxPackage" )
            typelist+=( "Joomla" "KodExplorer" "LogCenter" "MailClient" "MailPlus-Server" )
            typelist+=( "MailServer" "MailStation" "MariaDB" "MariaDB10" "MediaServer" )
            typelist+=( "MediaWiki" "MEGAcmd" "MigrationAssistant" "MinimServer" "Moodle" )
            typelist+=( "NBR" "NBR-Transporter" "nConnect" "Node.js_v12" "Node.js_v14" )
            typelist+=( "Node.js_v16" "Node.js_v18" "Node.js_v8" "NoteStation" )
            typelist+=( "NVIDIARuntimeLibrary" "OAuthService" "OrangeHRM" "osTicket" "PACS" )
            typelist+=( "PDFViewer" "Perl" "PetaSpace" "Phddns" "PhotoStation" "PHP5.6" "PHP7.0" )
            typelist+=( "PHP7.2" "PHP7.3" "PHP7.4" "PHP8.0" "PHP8.1" "PHP8.2" "phpMyAdmin" )
            typelist+=( "Plex Media Server" "PlexMediaServer" "PrestoServer" "ProxyServer" "py3k" )
            typelist+=( "Python" "Python2" "Python3.9" "PythonModule" "QuickConnect" "QuikFynd" )
            typelist+=( "RadiusServer" "RagicBuilder" "ReplicationService" "resiliosync" )
            typelist+=( "SafeAccess" "ScsiTarget" "SecureSignIn" "SMBService" "SnapshotReplication" )
            typelist+=( "Sony_BraviaSignage" "Spreadsheet" "SSOServer" "StorageAnalyzer" )
            typelist+=( "StorageManager" "SupportService" "SurveillanceDevicePack" )
            typelist+=( "SurveillanceLocalDisplay" "SurveillanceStation" "SVN" "syncthing.net" )
            typelist+=( "SynoFinder" "SynologyApplicationService" "SynologyDrive" )
            typelist+=( "SynologyFileManager" "SynologyMoments" "SynologyPhotos" )
            typelist+=( "SynoOnlinePack_v2" "SynoSmisProvider" "SyslogServer" "Tailscale" )
            typelist+=( "TeamViewer" "TextEditor" "ThreatPrevention" "Tomcat7" "UniversalViewer" )
            typelist+=( "USBCopy" "VideoStation" "VirtualHere" "Virtualization" "VPNCenter" )
            typelist+=( "VPNPlusServer" "vtigerCRM" "WebDAVServer" "WebStation" "WordPress" )
            break ;;
        Utility)
            typelist=( "All" "ActiveBackupBusinessAgent" "ActiveBackupforRecoveryTool" )
            typelist+=( "ActiveBackupRecoveryMediaCreator" "Assistant" )
            typelist+=( "AxisBarcodeToSynologySurveillanceStation" "BeeDrive" )
            typelist+=( "BeeDriveOnlineInstaller" "C2BackupAgent" "C2BackupRecoveryTool" )
            typelist+=( "C2BackupRestoreMediaCreator" "C2IdentityADSyncTool" "C2IdentityAgent" )
            typelist+=( "C2IdentityDirectoryExportTool" "C2IdentityDockerMgmtTool" )
            typelist+=( "C2IdentityLDAPSyncTool" "C2IdentityWindowsADExportTool" "ChatClient" )
            typelist+=( "CloudStationBackup" "CloudStationDrive" "EvidenceIntegrityAuthenticator" )
            typelist+=( "HybridShareDownloadTool" "HyperBackupExplorer" "iSCSIVssPlugin" )
            typelist+=( "NFSVAAIPlugin" "NoteStationClient" "PhotoStationUploader" "Presto" )
            typelist+=( "StorageConsoleforVMware" "StorageConsoleforWindows" )
            typelist+=( "SurveillanceStationClient" "SurveillanceVideoConverterTool" )
            typelist+=( "SynologyCameraTool" "SynologyCloudSyncDecryptionTool" )
            typelist+=( "SynologyDriveClient" "SynologyFileManagerClient" )
            typelist+=( "TransactionsCompatibilityTool" "TransactionsDeviceSimulator" )
            typelist+=( "VMwarePlugin" "VMwareSRA" "VPNPlusClient" "WinVirtioDriver" )
            break ;;
        Mobile)
            typelist=( "All" "Android-ActiveInsight" "Android-BeeDrive" "Android-Drive" )
            typelist+=( "Android-DSaudio" "Android-DScam" "Android-DSchat" "Android-DScloud" )
            typelist+=( "Android-DSdownload" "Android-DSfile" "Android-DSfinder" "Android-DSmail" ) 
            typelist+=( "Android-DSnote" "Android-DSphoto" "Android-DSrouter" "Android-DSvideo" )
            typelist+=( "Android-LiveCam" "Android-Moments" "Android-Photos" "Android-SecureSignIn" )
            typelist+=( "AndroidTV-DSphoto" "AndroidTV-DSvideo" "AndroidTV-SynologyPhotos" )
            typelist+=( "Android-VPNPlus" )
            break ;;
        ChromeApp)
            typelist=( "All" "NoteStationClipper" "OfficeExtension" )
            break ;;
        ToolChain)
            typelist=( "All" "Synology NAS GPL Source" "toolchain" "toolkit" )
            break ;;
        Firmware)
            #typelist=( "All" "Camera" )
            typelist=( "Camera" )
            break ;;
        *)
            echo -e "Invalid choice" ;;
    esac
done
echo -e "You selected: $srcdir \n"


PS3="Choose $srcdir type: "
select subdir in "${typelist[@]}"; do
    type="$subdir"
    break
done
echo -e "You selected: $type \n"


if [[ -n $srcdir ]] && [[ -n $type ]]; then
    if [[ -n $php_log ]]; then
        php "$php_script" "$srcdir" "$type" 2>&1 | tee "$php_log"
    else
        php "$php_script" "$srcdir" "$type"
    fi
else
    echo "args empty!"  # debug
    exit 1              # debug
fi



clone.php
<?php
#----------------------------------------------------------------------
# Forked from https://github.com/stopforumspam/download-synology-dsm
#----------------------------------------------------------------------
# https://github.com/007revad/Download_Synology_Archive
#
# Checked at https://phpcodechecker.com/
#
# Run command
# php ./clone.php Os DSM 2>&1 | tee ./clone_$(date '+%Y%m%d-%H%M').log
#----------------------------------------------------------------------

$destination = "/volume1/downloads/archive.synology.com";

if (($argc != "3") || ($argv[1] == "Firmware")) {
    if ($argc < "2") {
        echo "Missing arguments!" . "\n";
    }
    elseif ($argc < "3") {
        echo "Missing argument!" . "\n";
    }
    elseif ($argc > "3") {
        echo "Too many arguments!" . "\n";
    }

    if ($argv[1] == "Firmware") {
        echo "Firmware needs to include the firmware type in <scrdir>" . "\n";
    }

    $i = 1;
    $args = "$argv[0]";
    while($i < $argc) {
        $args = $args . " \"" . "$argv[$i]" . "\"";
        $i++;
    }
    echo $args . "\n\n";

    echo "Usage: clone.php <srcdir> <subdir>" . "\n";
    echo "  Examples: " . "\n";
    echo "    php ./clone.php Os DSM" . "\n";
    echo "    php ./clone.php Package Docker" . "\n";
    echo "    php ./clone.php ToolChain \"Synology NAS GPL Source\"" . "\n\n";
    echo "To download all in the <srcdir> use All as the <subdir>" . "\n";
    echo "  Examples: " . "\n";
    echo "    php ./clone.php Os All" . "\n";
    echo "    php ./clone.php Package All" . "\n\n";
    echo "Firmware needs to include the firmware type in <scrdir>" . "\n";
    echo "  Examples: " . "\n";
    echo "    php ./clone.php Firmware/Camera BC500" . "\n";
    echo "    php ./clone.php Firmware/Camera TC500" . "\n";
    echo "    php ./clone.php Firmware/Camera All" . "\n\n";
    exit(1);
} else {
    $srcdir = "$argv[1]";
    $type = "$argv[2]";
    echo "------------------------------------------------------------ \n";
    //echo "Downloading: " . $srcdir . " > " . $type . "\n\n";
}

if(@chdir($destination)===true) { 
    echo "Current directory: " . getcwd() . "\n";
    sleep(5);  # Give me time to cancel if current directory incorrect
} else {
    echo "Change directory failed: " . $destination . "\n";
    exit(1);
}
echo "------------------------------------------------------------ \n";

function getLinks($url, $srcdir, $type, $dir) {

    //echo "url: $url" . "\n";  # debug

    $html = file_get_contents($url);
    if (empty($html)) {
        echo "file_get_contents command timed out!" . "\n";
        return;
    }
    $dom = new DOMDocument;
    @$dom->loadHTML($html);
    foreach ($dom->getElementsByTagName('a') as $node) {

        if (strpos($node->getAttribute("href"), ".pat") !== false  // Os
            || strpos($node->getAttribute("href"), ".spk") !== false  // Package
            || strpos($node->getAttribute("href"), ".bz2") !== false  // Utility
            || strpos($node->getAttribute("href"), ".deb") !== false  // Utility
            || strpos($node->getAttribute("href"), ".dmg") !== false  // Utility
            || strpos($node->getAttribute("href"), ".eap") !== false  // Utility
            || strpos($node->getAttribute("href"), ".elf") !== false  // Utility
            || strpos($node->getAttribute("href"), ".exe") !== false  // Utility
            || strpos($node->getAttribute("href"), ".gz") !== false   // Utility
            || strpos($node->getAttribute("href"), ".iso") !== false  // Utility
            || strpos($node->getAttribute("href"), ".msi") !== false  // Utility
            || strpos($node->getAttribute("href"), ".ova") !== false  // Utility
            || strpos($node->getAttribute("href"), ".pkg") !== false  // Utility
            || strpos($node->getAttribute("href"), ".rpm") !== false  // Utility
            || strpos($node->getAttribute("href"), ".vib") !== false  // Utility
            || strpos($node->getAttribute("href"), ".zip") !== false  // Utility
            || strpos($node->getAttribute("href"), ".apk") !== false  // Mobile
            || strpos($node->getAttribute("href"), ".crx") !== false  // ChromeApp
            || strpos($node->getAttribute("href"), ".txz") !== false  // ToolChain
            || strpos($node->getAttribute("href"), ".tbz") !== false  // ToolChain
            || strpos($node->getAttribute("href"), ".tgz") !== false  // ToolChain
            || strpos($node->getAttribute("href"), ".bin") !== false  // Firmware (Camera)
        ) {

            //echo "FILE: " . $node->getAttribute("href")."\n" ; 

            $remote = parse_url($node->getAttribute("href"));
            $fullpath = explode("/", $remote["path"]);
            $filename = urldecode(array_pop($fullpath));
            $path = "download/$srcdir/$type/$dir/";

            @mkdir($path, 0777, true);
            $dest = "$path$filename";

            # Getting remote file size
            $ch = curl_init($node->getAttribute("href"));
            curl_setopt($ch, CURLOPT_NOBODY, true);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_HEADER, true);
            curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
            $data = curl_exec($ch);
            curl_close($ch);

            $contentLength = 'unknown';
            $status = 'unknown';
            if (preg_match('/^HTTP\/(1\.[01]|2) (\d\d\d)/', $data, $matches)) {
                 $status = (int)$matches[2];
            }
            if (preg_match('/Content-Length: (\d+)/i', $data, $matches)) {
                 $contentLength = (int)$matches[1];
            }

            if ($status == 200 && $contentLength <> "unknown" && file_exists($dest) && (filesize($dest) != $contentLength)) {
                echo "Remote file is $contentLength however local file is " . filesize($dest) . ", Redownloading\n";
                @unlink($dest);
            }

            if (!file_exists($dest)) {
                //echo "downloading " . $node->getAttribute("href") . " to $dest \n";
                echo "Downloading: " . "$dest \n";
                // echo "From: " . $node->getAttribute("href") . "\n\n";

                $fh = fopen($dest, "w");
                $ch = curl_init();
                curl_setopt($ch, CURLOPT_USERAGENT,"Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:109.0) Gecko/20100101 Firefox/113.0 ");             
                curl_setopt($ch, CURLOPT_URL, $node->getAttribute("href"));
                curl_setopt($ch, CURLOPT_FILE, $fh);
                curl_exec($ch);
                curl_close($ch); 
//            } else {
//                // echo "Skipping download\n";
//                echo "Skipping: " . "$dest \n";
            }
        } elseif (strpos($node->getAttribute("href"), "/download/$srcdir/$type/") !== false) {
            echo "DIR: " . $node->getAttribute("href")."\n";
            $url = "https://archive.synology.com" .$node->getAttribute("href");
            getLinks($url, $srcdir, $type, $node->nodeValue);
        }
    }
}

function getDirs($url, $srcdir, $type) {
    global $types;
    $types = array();

    $html = file_get_contents($url);
    $dom = new DOMDocument;
    @$dom->loadHTML($html);
    foreach ($dom->getElementsByTagName('a') as $node) {

        $remote = parse_url($node->getAttribute("href"));
        $fullpath = explode("/", $remote["path"]);
        $type = urldecode(array_pop($fullpath));

        if (($type != "download") && (!str_contains($srcdir, $type))) {
            $types[] = $type;
            $path = "download/$srcdir/$type/";
        }
    }
}


$start = date('Y-m-d H:i:s');

if ($type != "All") {
    echo "Checking: " . $srcdir . "/" . $type . "\n";
    getLinks("https://archive.synology.com/download/$srcdir/$type", "$srcdir", "$type", "");
} else {
    echo "Checking: " . $srcdir . " > " . $type . "\n";
    getDirs("https://archive.synology.com/download/$srcdir", $srcdir, "");

    # Get array of types
    $type = "";
    $i = 0;
    while($i < count($types)) {
        echo "\nChecking: " . $srcdir . "/" . $types[$i] . "\n";
        getLinks("https://archive.synology.com/download/$srcdir/$types[$i]", "$srcdir", "$types[$i]", "");
        $i++;
    }
}

echo "------------------------------------------------------------ \n";
echo "Started:  $start \n";
echo "Finished: " . date('Y-m-d H:i:s') . "\n";
echo "------------------------------------------------------------ \n";
 
